.github/workflows/publish-to-test-pypi.yml

```diff
@@ -32,12 +37,17 @@ jobs:
         .
 
     - name: Publish distribution 📦 to Test PyPI
+      if: github.event_name == 'push'
       uses: pypa/gh-action-pypi-publish@release/v1
       with:
         password: ${{ secrets.TEST_PYPI }}
         repository-url: https://test.pypi.org/legacy/
+        skip_existing: true
+
     - name: Publish distribution 📦 to PyPI
-      if: startsWith(github.ref, 'refs/tags')
+#      if: startsWith(github.ref, 'refs/tags')
```

Comment:
you can think about a tag as a state of your code at a particular point in time. it's kind of like a branch [but it's a point in time / state of your code at that time](https://www.atlassian.com/git/tutorials/inspecting-a-repository/git-tag). this i think might have worked if you added tags to the triggers at the top.

but i do think using releases is just cleaner overall particularly if these are things you're pushing to pypi.

Reply:
I agree, it makes more sense to me to associate PyPI publishing with releases, not tags. I still don't understand why PyPA recommends this approach, but I don't want to spend any more time on it.

Reply:
yes i'm not sure what the thinking was there either. i think this approach is a good one to just use releases. you could just delete this line.